### PR TITLE
1359 new picklist

### PR DESCRIPTION
**Pull request by corykon**
Rewrote picklist component to support templating, virtual scrolling, and lots of other stuff. Previous picklist component is still there but labeled as deprecated on the docs site.

If folks are around, I can try to schedule a code review meeting next week on Monday or Tuesday. Feel free to dive in now and shoot me some feedback.

picklist2

### CODE REVIEW

File: projects/cashmere/src/lib/picklist2/value-utils.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    '&': '&amp;',
+    '<': '&lt;',
+    '>': '&gt;',
+    '"': '&quot;',
+    '\'': '&#39;'
+};
+
+export function escapeHTML(string: string) {
+    return (string && hasUnescapedHTMLExp.test(string)) ?
+        string.replace(unescapedHTMLExp, chr => htmlEscapes[chr]) :
```
Comment (petermcclanahan) [inline projects/cashmere/src/lib/picklist2/value-utils.ts:13]: @atkulp had a good idea when we were doing something like this in our code... Let the browser do the escaping for you. 

```
document.createElement('p'); // Throw-away element, or keep to reuse, but not in DOM
p.innerText = "<div>testing>br>&blah";  // Setting inner text causes escaping
p.innerHTML; // Get that escaped text back
```

File: projects/cashmere/src/lib/picklist2/pick-option.component.ts [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    template: `<ng-content></ng-content>`
+})
+export class PickOptionComponent implements OnChanges, AfterViewChecked, OnDestroy {
+    /** Value for this option. */
+    @Input() value: any;
+    /** If true, the option cannot be selected. */
+    @Input() set disabled(value: boolean) { this._disabled = this._isDisabled(value); }
+    get disabled() { return this._disabled; }
+
+    readonly _stateChange$ = new Subject<{ value: any, disabled: boolean, label?: string }>();
```
Comment (petermcclanahan) [inline projects/cashmere/src/lib/picklist2/pick-option.component.ts:26]: It's only used a few times here, but my personal preference would be to create a separate type/interface for:
```
{
value: any,
disabled: boolean,
label?: string
}

File: projects/cashmere/src/lib/picklist2/pane/pick-pane.component.ts [hunk 3]
```diff
@@ -0,0 +1,472 @@
...
+    /** Highlight all items in the list */
+    selectAll() {
+        this.itemsList.selectAll();
+        this._cd.detectChanges();
+    }
+
+    /**
+     * Shift + click selects a range
+     * Ctrl + click adds to current selection
+     * Regular click clears current selection and selects anew
```
Comment (petermcclanahan) [inline projects/cashmere/src/lib/picklist2/pane/pick-pane.component.ts:253]: I don't remember if the current picklist does this, but that's very nice!

**Review by petermcclanahan — APPROVED** [review]
There were a lot of files here, and I'm probably not the best to give feedback, but Andrew asked me to the other day since we will probably use this picklist as soon as it's merged in. That said, I think you've done a fabulous job. I didn't follow everything you've done, but I looked through everything. I love all the options available, and I'm excited to use this!

File: projects/cashmere/src/lib/picklist2/value-utils.ts (hunk 1 above)
Comment (corykon) [reply projects/cashmere/src/lib/picklist2/value-utils.ts]: love it, I've implemented this

File: projects/cashmere/src/lib/picklist2/pick-option.component.ts (hunk 2 above)
Comment (corykon) [reply projects/cashmere/src/lib/picklist2/pick-option.component.ts]: Good ask, made this change.

File: projects/cashmere-examples/src/lib/picklist2-hc-pick-option/picklist2-hc-pick-option-example.component.html [hunk 4]
```diff
@@ -0,0 +1,19 @@
+<p>You can provide options to the picklist in a declarative way using the <code>&#x3C;hc-pick-option&#x3E;</code> component.</p>
+<br>
+
+<hc-picklist2 [(ngModel)]="selected">
+    <hc-pick-option *ngFor="let veggie of veggies" [value]="veggie.id" [disabled]="veggie.disabled">
+        <hc-icon fontSet="fa" fontIcon="fa-cutlery" hcIconSm></hc-icon>&nbsp;
+        {{veggie.name}}
+    </hc-pick-option>
+    <hc-pick-option [value]="21">Raddish</hc-pick-option>
```
Comment (jacquie) [inline projects/cashmere-examples/src/lib/picklist2-hc-pick-option/picklist2-hc-pick-option-example.component.html:9]: nit: typo: Radish only has one 'n'

**Review by andrew-frueh — APPROVED** [review]
Pending the few changes we talked about in the code review today, thumbs up from me.  Amazing work @corykon

File: projects/cashmere-examples/src/lib/picklist2-hc-pick-option/picklist2-hc-pick-option-example.component.html (hunk 4 above)
Comment (corykon) [reply projects/cashmere-examples/src/lib/picklist2-hc-pick-option/picklist2-hc-pick-option-example.component.html:9]: fixed :)

**Comment by andrew-frueh** [comment]
Looks like you've got one line length linting error @corykon - then good to merge

**Comment by corykon** [comment]
@andrew-frueh 
> Looks like you've got one line length linting error @corykon - then good to merge

Yep I'll fix that. Still working on the usage guide for migration and renaming `hc-picklist2` -> `hc-picklist`. Getting close.

**Pull request merged**